### PR TITLE
[ZEPPELIN-3182] Support saving notebooks to Google Cloud Storage

### PR DESCRIPTION
**Pull request by karth295**
### What is this PR for?

Support saving notebooks to Google Cloud Storage, similar to implementations for S3 and Azure. It uses the same authentication mechanisms as the BigQuery interpreter.

I am new to Maven, so please check my work on the pom.xml files. In particular, I upgraded Guava to 23.0, which was required for `google-cloud-java`. Going through hello-world with my changes seems to work.

Also, I modified the BigQuery interpreter docs to point to the **latest** GCS storage docs. Is it more appropriate to pin to the version you are viewing? How can I do that?

### What type of PR is it?
Improvement

### Todos
* [Low priority] Support encryption keys

I don't this is particularly important, at least for v1.

### How should this be tested?
* I added unit tests for the core functionality
* I manually tested the authentication instructions (but that could use a second pair of eyes)

### Questions:
* Does the licenses files need update?
  * No idea. `google-cloud-java` is Apache 2: https://github.com/GoogleCloudPlatform/google-cloud-java/blob/master/LICENSE
* Is there breaking changes for older versions?
  * Nope.
* Does this needs documentation?
  * Yes, and I tried to update the docs (but there are likely other things that need to be updated)

### CODE REVIEW

**Comment by karth295** [comment]
The TravisCI failure doesn't look right to me:

```
[WARNING] 
Dependency convergence error for com.google.guava:guava:15.0 paths to dependency are:
+-org.apache.zeppelin:zeppelin-zengine:0.8.0-SNAPSHOT
  +-org.apache.zeppelin:zeppelin-interpreter:0.8.0-20170802.043121-7
    +-com.google.guava:guava:15.0
and
+-org.apache.zeppelin:zeppelin-zengine:0.8.0-SNAPSHOT
  +-com.google.guava:guava:23.0-android
[WARNING] The requested profile "web-ci" could not be activated because it does not exist.
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:1.3.1:enforce (enforce) on project zeppelin-zengine: org.apache.maven.plugins.enforcer.DependencyConvergence failed with message:
[ERROR] Failed while enforcing releasability the error(s) are [
[ERROR] Dependency convergence error for com.google.guava:guava:15.0 paths to dependency are:
[ERROR] +-org.apache.zeppelin:zeppelin-zengine:0.8.0-SNAPSHOT
[ERROR]   +-org.apache.zeppelin:zeppelin-interpreter:0.8.0-20170802.043121-7
[ERROR]     +-com.google.guava:guava:15.0
[ERROR] and
[ERROR] +-org.apache.zeppelin:zeppelin-zengine:0.8.0-SNAPSHOT
[ERROR]   +-com.google.guava:guava:23.0-android
```

If you run `mvn clean package`, it should not use the cached `org.apache.zeppelin:zeppelin-interpreter:0.8.0-20170802.043121-7`, and the dependencies should be correct. (Though maven still scares/confuses me)

File: zeppelin-zengine/pom.xml [hunk 1]
```diff
@@ -166,7 +249,14 @@
     <dependency>
       <groupId>com.google.guava</groupId>
       <artifactId>guava</artifactId>
-      <version>15.0</version>
+      <!-- Android is a bit of a misnomer, it is for Java 7 -->
+      <version>23.0-android</version>
```
Comment (zjffdu) [inline zeppelin-zengine/pom.xml]: Any reason to upgrade guava ? Guava's compatibility across versions is very weak, it is very likely to cause side effect when upgrading guava.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GCSNotebookRepo.java [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
+    Note note;
+    try {
+      note = Note.fromJson(new String(contents, encoding));
+    } catch (JsonParseException jpe) {
+      throw new IOException(
+          "Could note parse as json " + blobId.toString() + jpe.getMessage(), jpe);
+    }
+
+    for (Paragraph p : note.getParagraphs()) {
```
Comment (zjffdu) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GCSNotebookRepo.java]: If this post-processing is common to all notebook repo implementation, then it should be not be here.

File: zeppelin-zengine/pom.xml (hunk 1 above)
Comment (karth295) [reply zeppelin-zengine/pom.xml]: Unfortunately, yes. `google-cloud-java` seems to depend on things in 23.0. When I tried excluding guava from google-cloud-java, I got method not found errors at runtime.

That being said, if all of zeppelin's tests pass without hitting class not found / method not found errors, is it okay to upgrade Guava?

**Comment by zjffdu** [comment]
@karth295 Please try to rebase your PR, I think the build issue you mentioned above has been fixed.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GCSNotebookRepo.java (hunk 2 above)
Comment (karth295) [reply zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GCSNotebookRepo.java]: Done -- I did that refactor as a separate commit.

**Comment by karth295** [comment]
These two commits are rebased over master.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GCSNotebookRepo.java (hunk 2 above)
Comment (karth295) [reply zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GCSNotebookRepo.java]: Hah but I forgot to delete this code.

**Comment by zjffdu** [comment]
CI still fails, could you retrigger it again ? Just close and reopen this PR.

File: zeppelin-zengine/pom.xml (hunk 1 above)
Comment (zjffdu) [reply zeppelin-zengine/pom.xml]: Apparently, many travis CI fails.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/NotebookRepo.java [hunk 3]
```diff
@@ -46,28 +49,50 @@
    * @return
    * @throws IOException
    */
-  @ZeppelinApi public Note get(String noteId, AuthenticationInfo subject) throws IOException;
+  @ZeppelinApi
+  public Note get(String noteId, AuthenticationInfo subject) throws IOException {
+    Note note = getRaw(noteId, subject);
```
Comment (zjffdu) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/NotebookRepo.java]: You can put these logic in Note#postProcessParagraphs, so that you don't need to change the interface signature and lots of other things

File: zeppelin-zengine/pom.xml (hunk 1 above)
Comment (karth295) [reply zeppelin-zengine/pom.xml]: There's just one issue (causing two tests to fail):

```
java.lang.NoSuchMethodError: com.google.common.base.Objects.toStringHelper(Ljava/lang/Object;)Lcom/google/common/base/Objects$ToStringHelper;
	at org.apache.hadoop.metrics2.lib.MetricsRegistry.toString(MetricsRegistry.java:406)
	at java.lang.String.valueOf(String.java:2994)
	at java.lang.StringBuilder.append(StringBuilder.java:131)
	at org.apache.hadoop.ipc.metrics.RpcMetrics.<init>(RpcMetrics.java:74)
	at org.apache.hadoop.ipc.metrics.RpcMetrics.create(RpcMetrics.java:80)
	at org.apache.hadoop.ipc.Server.<init>(Server.java:2218)
	at org.apache.hadoop.ipc.RPC$Server.<init>(RPC.java:951)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Server.<init>(ProtobufRpcEngine.java:534)
	at org.apache.hadoop.ipc.ProtobufRpcEngine.getServer(ProtobufRpcEngine.java:509)
	at org.apache.hadoop.ipc.RPC$Builder.build(RPC.java:796)
	at org.apache.hadoop.hdfs.server.namenode.NameNodeRpcServer.<init>(NameNodeRpcServer.java:351)
	at org.apache.hadoop.hdfs.server.namenode.NameNode.createRpcServer(NameNode.java:674)
	at org.apache.hadoop.hdfs.server.namenode.NameNode.initialize(NameNode.java:647)
	at org.apache.hadoop.hdfs.server.namenode.NameNode.<init>(NameNode.java:812)
	at org.apache.hadoop.hdfs.server.namenode.NameNode.<init>(NameNode.java:796)
	at org.apache.hadoop.hdfs.server.namenode.NameNode.createNameNode(NameNode.java:1493)
	at org.apache.hadoop.hdfs.MiniDFSCluster.createNameNode(MiniDFSCluster.java:1115)
	at org.apache.hadoop.hdfs.MiniDFSCluster.createNameNodesAndSetConf(MiniDFSCluster.java:986)
	at org.apache.hadoop.hdfs.MiniDFSCluster.initMiniDFSCluster(MiniDFSCluster.java:815)
	at org.apache.hadoop.hdfs.MiniDFSCluster.<init>(MiniDFSCluster.java:475)
	at org.apache.hadoop.hdfs.MiniDFSCluster$Builder.build(MiniDFSCluster.java:434)
	at org.apache.zeppelin.interpreter.MiniHadoopCluster.start(MiniHadoopCluster.java:41)
	at org.apache.zeppelin.interpreter.SparkInterpreterModeTest.setUp(SparkInterpreterModeTest.java:29)
```

This is very unfortunate. Hadoop's `trunk` and `branch-2` no longer use this (now deprecated) class: `https://github.com/apache/hadoop/blob/branch-2/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/metrics2/lib/MetricsRegistry.java`. Zeppelin depends on Hadoop 2.7.

Comment (karth295) [reply zeppelin-zengine/pom.xml]: The fix is in Hadoop 2.9.0 and Hadoop 2.8.3 https://issues.apache.org/jira/browse/HADOOP-14891

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/NotebookRepo.java [hunk 4]
```diff
@@ -67,7 +67,7 @@
   /**
    * Release any underlying resources
    */
-  @ZeppelinApi public void close();
+  @ZeppelinApi public abstract void close();
```
Comment (zjffdu) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/NotebookRepo.java]: abstract is not necessary

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/FileSystemNotebookRepo.java [hunk 5]
```diff
@@ -1,12 +1,10 @@
 package org.apache.zeppelin.notebook.repo;
 
-import org.apache.commons.lang.StringUtils;
```
Comment (zjffdu) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/FileSystemNotebookRepo.java]: I created https://issues.apache.org/jira/browse/ZEPPELIN-3137 for code sytle, it is better to do it in some dedicated PRs.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/NotebookRepo.java (hunk 4 above)
Comment (karth295) [reply zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/NotebookRepo.java]: Good catch -- I didn't realize `abstract` was valid syntax in an interface.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/FileSystemNotebookRepo.java (hunk 5 above)
Comment (karth295) [reply zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/FileSystemNotebookRepo.java]: Thanks -- I did a bad job cleaning up my previous commit that turned `NotebookRepo` into an abstract class. I amended the commit so that I didn't change imports in classes I didn't actually modify.

**Comment by karth295** [comment]
FYI I finally resolved the dependency issues by using guava 20.0, which is the only version that works for both the `google-cloud-java` libraries as well as Hadoop (the deprecated `Objects` class).

**Comment by karth295** [comment]
@zjffdu I noticed that you just bumped the version to 0.9 -- can this feature be integrated into branch-0.8 release as well?

**Comment by zjffdu** [comment]
@karth295 The only concern is guava, if that doens't affect other function, it is fine to me to merge into 0.8 as well

**Comment by karth295** [comment]
The new CI failure doesn't seem real to me -- going to close/reopen to make it run again.

**Comment by karth295** [comment]
@zjffdu do you know what's causing these failure? (I can't tell whether my PR is breaking the tests)

**Comment by zjffdu** [comment]
The failure should due to jar file corrupted, could you clean travis cache and trigger the failed build again ? (I mean only rerun the failed build, not all the travis build matrix)

**Comment by karth295** [comment]
Nice, it's green now (though Jenkins didn't update).

**Comment by zjffdu** [comment]
Thanks @karth295 will merge if no more comments

File: zeppelin-zengine/pom.xml [hunk 6]
```diff
@@ -39,6 +39,7 @@
     <hadoop.version>2.7.3</hadoop.version>
     <commons.lang3.version>3.4</commons.lang3.version>
     <commons.vfs2.version>2.0</commons.vfs2.version>
+    <gcs.storage.version>1.14.0</gcs.storage.version>
```
Comment (felixcheung) [inline zeppelin-zengine/pom.xml]: is this a new dependency? let's add that to the LICENSE file

File: zeppelin-zengine/pom.xml [hunk 7]
```diff
@@ -114,10 +116,84 @@
       <artifactId>httpasyncclient</artifactId>
     </dependency>
 
+    <dependency>
+      <groupId>com.google.cloud</groupId>
+      <artifactId>google-cloud-storage</artifactId>
+      <version>${gcs.storage.version}</version>
```
Comment (felixcheung) [inline zeppelin-zengine/pom.xml]: ditto any new dependencies in this group

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GCSNotebookRepo.java [hunk 8]
```diff
@@ -0,0 +1,217 @@
+package org.apache.zeppelin.notebook.repo;
```
Comment (felixcheung) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GCSNotebookRepo.java]: need to add Apache header to any new file

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/repo/GCSNotebookRepoTest.java [hunk 9]
```diff
@@ -0,0 +1,218 @@
+package org.apache.zeppelin.notebook.repo;
```
Comment (felixcheung) [inline zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/repo/GCSNotebookRepoTest.java]: Need Apache header

File: zeppelin-zengine/pom.xml (hunk 6 above)
Comment (karth295) [reply zeppelin-zengine/pom.xml]: Done

File: zeppelin-zengine/pom.xml (hunk 7 above)
Comment (karth295) [reply zeppelin-zengine/pom.xml]: Done

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GCSNotebookRepo.java (hunk 8 above)
Comment (karth295) [reply zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GCSNotebookRepo.java]: Done

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/repo/GCSNotebookRepoTest.java (hunk 9 above)
Comment (karth295) [reply zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/repo/GCSNotebookRepoTest.java]: Done
